[14.0][IMP] payroll: settings, code not required, payslip batch with company

norlinhenrik opened the pull request:
See PR #72 and the discussion in issue #92

**Multi-company**: I have added hr_payslip_run.company_id

**Settings**: I have added a setting for "details by salary rule category". I have also rewritten some descriptions.

**Codes**: I have set `required=False` on

* hr_salary_rule.code
* hr_salary_rule.category
* hr_salary_rule_category.code
* hr_payroll_structure.code

@mtelahun @nimarosa @pedrobaeza Is this ok with you?
Existing tests are passing, and a manual test also worked well.

norlinhenrik committed:
[IMP] payroll: hr.payslip.run with company_id

[FIX] payroll: fix views

norlinhenrik committed:
[IMP] payroll: code & category are not required

norlinhenrik committed:
[IMP] payroll: settings descriptions, show_details_by_salary_rule_category

nimarosa commented:
I think codes should continue being required. Without codes you can't access rules and categories values from salary rules. 

Also if you remove the codes, when payroll construct the localdict
you will get errors because localdict need codes. Test are passing because all tests use codes, if you make a rule without codes they will fail. 

The company_id field is okay, since the other PR was not finished. 

The setting I have another idea, if we create a new on/off flag for in dynamic_filtered_lines like we do with child rules but for appears on pauslip, we can remove details_by_salary_rule_category directly.

mtelahun commented:
On 06/11/2022 16:57, Henrik Norlin wrote:
> *Codes*: I have set |required=False| on
>
>   * hr_salary_rule.code
>   * hr_salary_rule.category
>   * hr_salary_rule_category.code
>   * hr_payroll_structure.code
>
> Message ID: ***@***.***> 

Hi Henrik, while I am not against these changes I think that in practice 
hr_salary_rule.code and hr_salary_rule_category.code should probably 
have required="1" in the view because writing any meaningful salary 
rules that do not not reference them is not possible at the moment.

norlinhenrik committed:
[IMP] payroll: required fields, removed details_by_salary_rule_category

norlinhenrik commented:
Thanks for feedback. These fields are now required in the view:

   * hr_salary_rule.code
   * hr_salary_rule.category
   * hr_salary_rule_category.code

In payslip form view, I removed details_by_salary_rule_category and added `appears_on_payslip` to the payslip lines.

nimarosa commented:
Note that line_ids are not used anymore in the form. The main table of the form comes from the field dynamic_filtered_line_ids which is a computed field that checks for a Boolean field to show/hide child rules from the table.

What I suggested is adding a new Boolean troggle field for filtering this computed field and show/hide lines that appears on payslip, the same logic we are currently doing with child rules.

This will allow to show/hide the lines in this dynamic table so details_by_salary_rule_category can be removed. 

line_ids should not have any filter or limitation since it's the main object and should always return all the lines. 

Now the dynamic filtered payslip lines object should be changed by context to be used in reports / account moves from payroll account module. 

I hope I made it clear, if not, you can leave this change out of this PR and I can make one with this change and removing details by salary rule category. 

---

Finally I think that making the category and code fields required in view is not enough since this will allow to create rules without code from other places like xmlrpc, xml files, or database level insertions. The required=True in the model is the best approach, we should leave that like it is.

norlinhenrik committed:
[IMP] payroll: hide_invisible_lines (who don't appear on payslip)

mtelahun commented:
On 07/11/2022 01:24, Nicolas Rodriguez Sande wrote:
> Finally I think that making the the code fields required in view is 
> not enough since this will allow to create rules without code from 
> other places like xmlrpc or database level insertions. The 
> required=True in the model is the best approach, we should leave that 
> like it is.

For my part removing required= from field definitions is OK for me 
because there really isn't any technical reason for it and you could, 
conceivably, create rules without entering anything in any of these 
fields. I will admit that whether such rules would be any use is 
debatable. There is also a minor benefit as well: writing tests is 
faster when there are fewer "required" fields because when creating 
records you spend less time entering fields that you may not be 
interested in, but are required by the model.

I support adding required="1" in the views because it is a helpful 
visual cue for the user. Also, I am not sure but I think I have noticed 
instances in the Odoo CE addons where they are removing "required" from 
the model definitions and adding them to the views instead. I could be 
wrong though.


Regards,

Mike.

norlinhenrik commented:
I have added a toggle "Hide Invisible Lines" (not appearing on payslip).

`rule.code` makes it possible for one rule to override the values computed by a previous rule. **When is this useful?** 
**What else** may the code be useful for?
When a payroll user makes a **custom rule**, why bother to create a code? I don't use the code for anything!

nimarosa commented:
@mtelahun the problem about removing it from the model is that if you create a rule without code localdict generation will fail since it's using code as the key to generate the dict. The same happen with categories, sum of the rule categories will fail. 

So I continue thinking that for now, if we don't come up with a use case when you really can use a rule without code and we adapt the localdict code to that before the change: we should leave the required=True in model. 

If you want to test what I'm saying create some rules and categories without code and try to compute payslip.

nimarosa commented:
> I have added a toggle "Hide Invisible Lines" (not appearing on payslip).
> 
> 
> 
> `rule.code` makes it possible for one rule to override the values computed by a previous rule. **When is this useful?** 
> 
> **What else** may the code be useful for?
> 
> When a payroll user makes a **custom rule**, why bother to create a code? I don't use the code for anything!
> 

@norlinhenrik 
Rule code it's not for overriding anything. Rule code is for retrieving the total value of previously computed rule. To give you an example:

You need to write a salary rule, that calculates the 10% of the total value of another. How do you do this? 

Suposing the first salary rule has the code BASIC, you do:

result = BASIC * 0.10 

You are not overriding anything you just are retrieving BASIC salary rule value. 

I think the problem is that you don't understand what codes are for, maybe you never have to do a payslip with uses values from another lines but it's very common in payroll systems. 

If you want to look it at the code, all rules and totals gets added to the base of the localdict dict, that is what allows us to access them only typing the code. Maybe we should add some documentation about it since I think you are not the only person who didn't know it. Previously computed rules values can be accessed just by typing the code, it can be in the Python mode but also in the other modes.  

The same is with salary rule category codes, allows to access the accumulated value of rules that have that category.

norlinhenrik commented:
> If you want to test what I'm saying create some rules and categories without code and try to compute payslip.

I did. It works very well for me.

> maybe you never have to do a payslip with uses values from another lines

That's right. I have carefully assigned rules into categories, and for payslip totals I only use category codes, not rule codes. For tax reporting I register custom info on each rule, and I am going to create hr_custom_info depending on base_custom_info. Custom info cannot fit into a code. So currently I don't see any value of rule codes.

> Maybe we should add some documentation

Yes documentation is needed for new people to understand the advanced calculation options. I did know, though, that rule codes may be used the way you just described. (But I forgot about it as I was writing the previous message.)

> Rule code it's not for overriding anything.

Create Rule1 and Rule2 and:
* Set the same code on both of them. Then one rule will override the other rule. A user would not expect this.
* Remove the code on both of them. Then localdict will use the rule.**id** (see [here](https://github.com/OCA/payroll/pull/93/files#diff-d76b4c61a197708c60a662f32c7de7b7645e1a83565a0a0a6793d60907068651R543)) and show both lines. I prefer this approach.

nimarosa commented:
> I did. It works very well for me.

Okay so if you create a rule without code how do you access it from inside salary rules? If I want to use your new salary rule without code, how do I access it like a variable? 

> That's right. I have carefully assigned rules into categories, and for payslip totals I only use category codes, not rule codes. For tax reporting I register custom info on each rule, and I am going to create hr_custom_info depending on base_custom_info. Custom info cannot fit into a code. So currently I don't see any value of rule codes.

Okay so you use categories for everything. That's okay anyone can choose to compute payslips the way they want. But if I only need the value of one rule, why I'm going to create a category if I can just retrieve rule value using the code? 

I also store custom info for tax reporting, that's something that change in each country so it's okay we don't have a general solution for that. That belongs more to a localization module. 

Rule codes are not meant to store tax information or anything like that. They allow you to access rule values using the code in other salary rules. That's the use, and it's a very useful one. Maybe you don't use it which it's okay but I think you are trying to see codes like other thing that which they are. 

The "value" of rules codes it's just allowing you to access previously computed values from another rule. Anything more anything less. 


> Yes documentation is needed for new people to understand the advanced calculation options. I did know, though, that rule codes may be used the way you just described.

Yes I'll add this in some of my next PR. 

> Create Rule1 and Rule2 and:
> 
> * Set the same code on both of them. Then one rule will override the other rule. A user would not expect this.

I know about this behavior. What I was saying is this is not the main use of rule codes. The main use is to allow accessing the rules.

Anyways, this override behavior is on purpose, this allow you to make progressive salary rules. An example of this can be: 

You have a rule that needs to be recomputed more than one time, depending of lines. So you can create a rule that uses the value of the previously computed rule (using the code to get it) and sum another value.

This is also useful when you have to change salary rules in external modules. So instead of changing the original rule you just create another rule with the same code and so your logic there using the previously
Computed value of the original. So you don't have to change it. 

> * Remove the code on both of them. Then localdict will use the rule.**id** (see [here](https://github.com/OCA/payroll/pull/93/files#diff-d76b4c61a197708c60a662f32c7de7b7645e1a83565a0a0a6793d60907068651R543)) and show both lines. I prefer this approach.

Yes but using the ID it's not useful. The user don't know rule ID. 

--- 

I can understand that you don't use codes, and it limit you having to create them but I think this functionality is out of scope of the module, maybe you should add that to your custom_info module. Because on payroll it's on purpose that rules must have a code and that makes them a required field.

mtelahun commented:
Hi @nimarosa,

I'm not trying to defend creating rules without a code. I'm not even 
saying it should work. But from a technical standpoint making it a 
requirement in the model is not necessary. It is sufficient to make it 
required in the view. That also seems to be the way Odoo S.A is going.

On 07/11/2022 13:22, Nicolas Rodriguez Sande wrote:
>
> @mtelahun <https://github.com/mtelahun> the problem about removing it 
> from the model is that if you create a rule without code localdict 
> generation will fails since it's using code as the key to generate the 
> dict. The same happen with categories, sum of the rule categories will 
> fail.
>
> So I continue thinking that for now, if we don't come up with a use 
> case when you really can use a rule without code and we adapt the 
> localdict code to that before the change: we should leave the 
> required=True in model.
>
> If you want to test what I'm saying create some rules and categories 
> without code and try to compute payslip.
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/OCA/payroll/pull/93#issuecomment-1305394010>, or 
> unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AB4YOHESNAGFE2XUREHLCGDWHDJ5JANCNFSM6AAAAAARYOB3RI>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>

nimarosa commented:
> Hi @nimarosa,
> 
> I'm not trying to defend creating rules without a code. I'm not even 
> saying it should work. But from a technical standpoint making it a 
> requirement in the model is not necessary. It is sufficient to make it 
> required in the view. That also seems to be the way Odoo S.A is going.

Hi @mtelahun, yes I know what you saying. The only reason I believe it's better to have it in the model is to enforce codes in xml file creation of rules and in xmlrpc scripts. But if you guys believe it's okay to have it only in the view and this is the new odoo standard, we can leave it in the view. No problem with that.

mtelahun commented:
> Hi @nimarosa, I'm not trying to defend creating rules without a code. I'm not even saying it should work. But from a technical standpoint making it a requirement in the model is not necessary. It is sufficient to make it required in the view. That also seems to be the way Odoo S.A is going.
> […](#)
> On 07/11/2022 13:22, Nicolas Rodriguez Sande wrote: @mtelahun <https://github.com/mtelahun> the problem about removing it from the model is that if you create a rule without code localdict generation will fails since it's using code as the key to generate the dict. The same happen with categories, sum of the rule categories will fail. So I continue thinking that for now, if we don't come up with a use case when you really can use a rule without code and we adapt the localdict code to that before the change: we should leave the required=True in model. If you want to test what I'm saying create some rules and categories without code and try to compute payslip. — Reply to this email directly, view it on GitHub <[#93 (comment)](https://github.com/OCA/payroll/pull/93#issuecomment-1305394010)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB4YOHESNAGFE2XUREHLCGDWHDJ5JANCNFSM6AAAAAARYOB3RI>. You are receiving this because you were mentioned.Message ID: ***@***.***>

I posted this before reading the reply from @appstogrow . So, I'm glad to see that this actually works and that it is useful. In that case I withdraw my request to have it required in the view. I think his use-case is ample evidence that rule.code should NOT be a required field.  

@nimarosa My own view was much like yours: I didn't see how you could create a useful payroll system without rule codes. My reason for not objecting was that the rule code is not required for any technical reason. But if Henrik has managed to create a payroll system in which he does not use rule codes, then it is self-evidently obvious that they should not be required.

mtelahun commented on payroll/models/hr_payslip.py:
```diff
@@ -551,7 +540,7 @@ def _compute_payslip_line(self, rule, localdict, lines_dict):
         # compute the rule to get some values for the payslip line
         values = rule._compute_rule(localdict)
-        key = rule.code + "-" + str(localdict["contract"].id)
+        key = (rule.code or str(rule.id)) + "-" + str(localdict["contract"].id)
```
@nimarosa If the rule has a code it is used. Otherwise it uses the id. I think this is reasonable.

nimarosa replied:
It's acceptable. Maybe we should do the same for categories since what happens if categories don't have a code as required, we should use category id in the localdict? 

We should test this use cases at least one test to be sure everything is okay. 

@norlinhenrik @mtelahun

mtelahun commented:
@apps2grow Please add some test cases. I think most people will be using rule codes so if there are no test cases this is the type of thing where regressions can easily creep in unnoticed.

nimarosa commented:
> I posted this before reading the reply from @appstogrow . So, I'm glad to see that this actually works and that it is useful. In that case I withdraw my request to have it required in the view. I think his use-case is ample evidence that rule.code should NOT be a required field.  

@mtelahun if we do this, we should do it too to categories or leave categories code as required. 

I really don't see the use of not having a code but if you two think is useful, we can do it. 

> @nimarosa My own view was much like yours: I didn't see how you could create a useful payroll system without rule codes. My reason for not objecting was that the rule code is not required for any technical reason. But if Henrik has managed to create a payroll system in which he does not use rule codes, then it is self-evidently obvious that they should not be required.

I agree. For me codes are mandatory in almost all cases but no problem in leaving them not required if we take care of those cases. 

It's important to write test cases in this before doing it. 

--- 

@norlinhenrik @mtelahun maybe a good approach could be create a setting for enabling required in views? So people can choose the behavior. I suggest this thinking on the end user that might not know, also payroll manager could want codes to be required even if they are not necessary. 

If we make a setting that then we can check in views to apply or not the required attribute I think it could be a better solution that gives support for the two types of use cases.

norlinhenrik committed:
[IMP] payroll: New setting: Require code and category

norlinhenrik committed:
[IMP] payroll: test rule code in payslip line

norlinhenrik commented:
I have created a setting to require code and category.

For testing, I added `self.assertEqual(line.code, "TEST", "The code is 'TEST'")`. I cannot think of anything else to test.

nimarosa commented:
> I have created a setting to require code and category.
> 
> For testing, I added `self.assertEqual(line.code, "TEST", "The code is 'TEST'")`. I cannot think of anything else to test.

@norlinhenrik 
This is the problem if we use view required attribute we can't test the setting. But I think it's okay I will make que code review later. 

Other thing we should test is, when the rule has no code, we should test localdict values. Two cases, if has code, localdict should have a key which is rule code; if no code, ID should be the key in localdict. 

The same test has to happen with categories: categories without code, or categories with code. This is if you finally make code in category not required, if you leave it as required (didn't see code yet) you can skip categories test. 

Will review code when I get home so we can move forward.

nimarosa commented on payroll/models/hr_payroll_structure.py:
```diff
@@ -20,7 +20,7 @@ def _get_parent(self):
 
     name = fields.Char(required=True)
-    code = fields.Char(string="Reference", required=True)
+    code = fields.Char(string="Reference", required=False)
```
We should remove the required key instead of setting it to false. Because default value is false if not informed.

norlinhenrik replied:
Done

nimarosa commented on payroll/models/hr_payroll_structure.py:
```diff
@@ -42,6 +42,20 @@ def _get_parent(self):
+            .get_param("payroll.require_code_and_category")
+        )
+        self.require_code = require
+        return require
```
return is not required since its a compute method

norlinhenrik replied:
**compute** existing records needs `self.require_code = require`
**default** for new records needs `return require`

nimarosa replied:
You are right.

nimarosa commented on payroll/models/hr_payslip.py:
```diff
@@ -169,47 +164,41 @@ class HrPayslip(models.Model):
         compute="_compute_payslip_count", string="Payslip Computation Details"
     )
     hide_child_lines = fields.Boolean(string="Hide Child Lines", default=False)
+    hide_invisible_lines = fields.Boolean(string="Hide Invisible Lines", default=False)
```
please change the string to "Show only lines that appear on payslip" so it's more clear to the user.

norlinhenrik replied:
Done

nimarosa commented on payroll/models/hr_salary_rule.py:
```diff
@@ -12,7 +12,7 @@ class HrSalaryRule(models.Model):
     name = fields.Char(required=True, translate=True)
     code = fields.Char(
-        required=True,
+        required=False,
```
We should remove the required key instead of setting it to false. Because default value is false if not informed.

norlinhenrik replied:
Done

nimarosa commented on payroll/models/hr_salary_rule.py:
```diff
@@ -27,7 +27,7 @@ class HrSalaryRule(models.Model):
     )
     category_id = fields.Many2one(
-        "hr.salary.rule.category", string="Category", required=True
+        "hr.salary.rule.category", string="Category", required=False
```
We should remove the required key instead of setting it to false. Because default value is false if not informed.

norlinhenrik replied:
Done

nimarosa commented on payroll/models/hr_salary_rule_category.py:
```diff
@@ -9,7 +9,7 @@ class HrSalaryRuleCategory(models.Model):
 
     name = fields.Char(required=True, translate=True)
-    code = fields.Char(required=True)
+    code = fields.Char(required=False)
```
We should remove the required key instead of setting it to false. Because default value is false if not informed.

norlinhenrik replied:
Done

nimarosa commented on payroll/models/res_config_settings.py:
```diff
@@ -10,23 +10,28 @@ class ResConfigSettings(models.TransientModel):
+    )
+    require_code_and_category = fields.Boolean(
+        config_parameter="payroll.require_code_and_category",
+        string="Require code and category",
```
We can change the string to: "Require codes and categories when creating payroll objects"
And the help to: "If enabled, codes and categories are required when you create payroll objects that contain this fields" 

Or something more explanatory to a not technical user, suggestions accepted.

norlinhenrik replied:
Good that you think of text formulations.
The fields are required always, not just on creation. I've updated the text, but differently than you suggested.
No other help description say "If enabled". The sentences are usually "Do this", and you enable it if you want it.

norlinhenrik committed:
[IMP] payroll: removed required=False, improved text descriptions

nimarosa reviewed (changes requested):
@norlinhenrik Everything look good now, but i think you are missing a possible problem. See method _sum_salary_rule_category in hr_payslip.py. 

When we sum categories, we use the code to add the sum to localdict. Since now we are allowing categories without codes, this method will fail since no code will be found. You have two options: 

1. Leave categories codes as required always
2. Apply some logic like you've done in rule "key" and use category id if not code found. 

Personally i don't think categories have any use if we don't have the codes, but no problem if you want to maintain this approach, but we need to fix that method and maybe test it, because test are not failing but this method will fail if we have a category without code.

norlinhenrik commented:
How can we check the localdict values in a test?

nimarosa commented:
> How can we check the localdict values in a test?

@norlinhenrik Maybe using the localdict_hook which return the final localdict after computing a payslip. But it's just a idea. Other way will be creating rules that return a category value using code to access it and then one case using a category value with category_id for access.

But if testing it's difficult we can make the change and get it merged with a code and functional review. The most important is to cover the case so we don't have errors when someones uses a category without code.

norlinhenrik commented:
I have made a test now to make sure that code works as expected; and no code also works.

> When we sum categories, we use the code to add the sum to localdict. Since now we are allowing categories without codes, this method will fail since no code will be found.

A dict accepts None as a key. So it doesn't fail even with no code.

nimarosa commented:
> I have made a test now to make sure that code works as expected; and no code also works.
> 
> 
> 
> > When we sum categories, we use the code to add the sum to localdict. Since now we are allowing categories without codes, this method will fail since no code will be found.
> 
> 
> 
> A dict accepts None as a key. So it doesn't fail even with no code.
> 

@norlinhenrik 

So if this is the case I think we cannot let categories have no codes. Because if an empty dict is accepted as a key, more than one category will sum to the same dict object, so we are going to have strange behaviors. 

So we have to:

1. Ensure a category has at least one code or ID used as key in localdict
2. Ensure that a category can't have and empty dict key in localdict 

If we cannot garantice that, we can't allow categories have no codes, because they will loose their main function which is accumulate values.

nimarosa commented:
@norlinhenrik did you check if this is not also happening with rules?

norlinhenrik committed:
[IMP] payroll: test rule and category with and without code

norlinhenrik commented:
I don't see how `localdict["categories"][None]` could cause strange behavior. Anyway, I have added a condition:
```
        if category.code:
            localdict["categories"].dict[category.code] = (
                localdict["categories"].dict.get(category.code, 0) + amount
            )
```

norlinhenrik commented:
> @norlinhenrik did you check if this is not also happening with rules?

What do you mean?

nimarosa commented:
> I don't see how `localdict["categories"][None]` could cause strange behavior. Anyway, I have added a condition:
> 
> ```
> 
>         if category.code:
> 
>             localdict["categories"].dict[category.code] = (
> 
>                 localdict["categories"].dict.get(category.code, 0) + amount
> 
>             )
> 
> ```

What I mean is, let say we have two categories without code (different ones but both without code) when sum function run, both will return None as code, so we end up having two **different** categories summing to the same dict object. This is what I think it can't happen.

nimarosa commented:
> > @norlinhenrik did you check if this is not also happening with rules?
> 
> 
> 
> What do you mean?

@norlinhenrik the same as my last comment but with rules. If none is accepted as code, will the rule ever get ID assigned as code? Because if not, the same will happen and that will make rules to override each others.

norlinhenrik commented:
The only way to get the rule value or the category sum is by code.
When the rule value / category sum is **not needed in localdict**, the rule/category code may be empty.
Then I don't care whether `localdict["categories"][None]` exists or not. With the previous change, it will not exist.
With the change below, `localdict[None]` will not exist either.
```
        if rule.code:
            localdict[rule.code] = total
            localdict["rules"].dict[rule.code] = rule
            localdict["result_rules"].dict[rule.code] = BaseBrowsableObject(values)
```

nimarosa reviewed (approved):
Okay it's acceptable to me. So rules and categories without code, will not be included in localdict. 

Not a fan, but acceptable because there is a setting to turn this off. Merging soon.

nimarosa commented:
/ocabot rebase

nimarosa commented:
@norlinhenrik can you rebase this on top 14.0? Bot seems not want to work today. 

Edit: Forget it, it worked.

nimarosa commented:
/ocabot rebase

nimarosa commented:
/ocabot merge minor
